typing/typecore.ml

```diff
@@ -998,34 +998,43 @@ let check_scope_escape loc env level ty =
   with Unify trace ->
     raise(Error(loc, env, Pattern_type_clash(trace, None)))
 
-(* type_pat propagates the expected type as well as maps for
-   constructors and labels.
-   Unification may update the typing environment. *)
-(* constrs <> None => called from parmatch: backtrack on or-patterns
-   explode > 0 => explode Ppat_any for gadts *)
-let rec type_pat ?(exception_allowed=false) ~constrs ~labels ~no_existentials
-          ~mode ~explode ~env sp expected_ty k =
+type pattern_typing_params = {
+  exception_allowed: bool;
+
+  (* maps for constructors and labels *)
+  (* constrs <> None => called from parmatch: backtrack on or-patterns *)
+  constrs: (string, Types.constructor_description) Hashtbl.t option;
+  labels: (string, Types.label_description) Hashtbl.t option;
+
+  no_existentials: existential_restriction option;
+  mode: type_pat_mode;
+
+  (* explode > 0 => explode Ppat_any for gadts *)
+  explode: int;
+}
+
+let rec type_pat params ~env sp expected_ty k =
   Builtin_attributes.warning_scope sp.ppat_attributes
     (fun () ->
-       type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
-         ~explode ~env sp expected_ty k
+       type_pat_aux params ~env sp expected_ty k
     )
 
-and type_pat_aux ~exception_allowed ~constrs ~labels ~no_existentials ~mode
-      ~explode ~env sp expected_ty k =
-  let mode' = if mode = Splitting_or then Normal else mode in
-  let type_pat ?(exception_allowed=false) ?(constrs=constrs) ?(labels=labels)
-        ?(mode=mode') ?(explode=explode) ?(env=env) =
-    type_pat ~exception_allowed ~constrs ~labels ~no_existentials ~mode ~explode
-      ~env
-  in
+and type_pat_aux params ~env sp expected_ty k =
+  let params =
```

Comment:
This shadows `params` whereas the previous version wasn't shadowing `mode`.
Seems like a bug.

Reply:
This logic was and still is a bit tricky, but I believe the change is correct.

Before, if you didn't use an explicit `~mode` parameter when calling the inner `type_pat` helper, you would get `mode'` as the mode. After the change, you get the shadowed-`params`, which is defined with the same modified `mode'` value. On the other hand, all call-sites that did specify an explicit `~mode` parameter now use `{ params with mode = ... }`, which (even with the post-shadowing `params` value) is correct.

Reply:
P.S.: @trefis if you are convinced by this explanation, could you please dismiss your "Changes requested" review? I would like to merge this soon to make the #8970 diff simpler.

Reply:
I think @trefis is right: once the value of `param` has been shadowed, `param.mode` is never equal to `Splitting_or`, which is problematic in the `Ppat_or` case at line 1363.

Reply:
That sounds convincing, indeed. It is interesting to note that none of the typing tests are failing.

Reply:
I updated the PR to fix this (I think), but I'm going to close it anyway to go with the weak-sauce solution proposed by Jacques.

Reply:
FTR: I agree with @Octachron, that's what I had in mind.

@gasche: do you fancy making another PR to add a test that would have broken with your initial change?

Reply:
I would rather not. I'm not familiar enough with this backtracking stuff to find a test easily, and I have other things on my TODO list that make me wish someone more versed in the problem would take care of it.

Reply:
I was thinking of taking a look at this testsuite issue.

Reply:
Note after further investigation, the previous bug was only a performance bug:

`Splitting_or` is used to share the information that all branches of a nested-or pattern `(A|B)|(C|D)` needs backtracking. Consequently, if it is set, the whole or-pattern should be split.

By resetting `Splitting_or` to `Normal` too early, we lose this shared information and need to recheck that all subranches needs backtracking after every split, thus adding an O(n^2) term to the time complexity.

I don't think it makes sense to try to add a performance test to the test suite. I would be instead more tempted to replace `Splitting_or` by an `split_or_until` iterator.